performance_manager/lib/l1_full_trip_events.py

```diff
@@ -0,0 +1,389 @@
+        result = db_manager.execute_with_data(
+            update_query,
+            insert_dataframe.loc[
+                tu_stopped_mask, ["hash", "fk_tu_stopped_event"]
+            ].rename(columns={"hash": "b_hash"}),
+        )
+        logging.info(
+            "%d 'fk_tu_stopped_event' values updated in FullTripEvents",
+            result.rowcount,
+        )
+
+
+def insert_new_events(db_manager: DatabaseManager) -> None:
+    """
+    insert new events into FullTripEvents table with INSERT from SELECT operation
+    table to table insertion
+    """
+    insert_cols = [
+        "hash",
+        "fk_vp_stopped_event",
+        "fk_vp_moving_event",
+        "fk_tu_stopped_event",
+    ]
+
+    insert_query = (
+        FullTripEvents.metadata.tables[FullTripEvents.__tablename__]
+        .insert()
+        .from_select(
+            insert_cols,
+            sa.select(
+                TempFullTripEvents.hash,
+                TempFullTripEvents.fk_vp_stopped_event,
+                TempFullTripEvents.fk_vp_moving_event,
+                TempFullTripEvents.fk_tu_stopped_event,
+            ).where(
+                (TempFullTripEvents.hash.notin_(sa.select(FullTripEvents.hash)))
+            ),
+        )
+    )
+    result = db_manager.execute(insert_query)
+    logging.info("%d rows inserted into FullTripEvents", result.rowcount)
+
+
+def process_full_trip_events(db_manager: DatabaseManager) -> None:
+    """
+    process new events from L0 tables and insert to L1 FullTripEvents RDS table
+
+    primary key and category columns will be selected from 3 data sources:
+        - RT_VEHICLE_POSITIONS (moving events)
+        - RT_VEHICLE_POSITIONS (stopped events)
+        - TRIP_UPDATES (stopped events)
+
+    primary keys from l0 tables are selected if they do not exist in their
+    corresponding foreign key column in FullTripEvents table
+
+    TRIP_UPDATE events are predictions used to backfill stop events for moving
+    events that had no associated stopped event (mostly commuter rail)
+
+    category columns from the 3 sources are hashed in the same manner and
+    events are matched together based on hashes, these hashes will function
+    as the primary key in the FullTripEvents table
+
+    all valid RT_VEHICLE_POSITIONS (moving & stopped) events are merged to create
+    records in the FullTRipEvents table (using outer join), so stop events may
+    exist without moving events and vice versa
+
+    TRIP_UPDATE stop events are only merged to a valid RT_VEHICLE_POSITIONS
+    moving event (using left join)
+
+    all merged new FullTripEvents records are inserted into temporary loading
+    table (TempFullTripEvents)
+
+    if computed hashes in TempFullTripEvents already exist in FullTripEvents
+    table, foreign key fields are updated with non-null key values
+
+    if computed hahes in TempFullTripEvents DO NOT exist in FullTripEvents,
+    full records are inserted from TempFullTripEvents into FullTripEvents
+    """
```

Comment:
Lots of new comments and explanations of business logic

Reply:
thanks for doing this.